### ISSUE TITLE
Clean URLs

### ISSUE THREAD

**Issue by shreyasminocha** [rust-lang/mdBook#1239]
I'm interested in clean URLs for pages: 

File: `src/foo/bar.md`
Expected path: `/foo/bar`
Actual path: `/foo/bar.html`

It's mostly a stylistic preference.

I'm transitioning from GitBook and this is one of the few small things I'm having trouble with. Thanks!

Related: #1238

**Comment by nihaals** [comment]
I think you're meant to use `README.md` to `index.html`

**Comment by shreyasminocha** [comment]
While I could create `src/foo/bar/readme.md` and access it at `/foo/bar`, it would be a lot more convenient to be accessible `/foo/bar`.

**Comment by shreyasminocha** [comment]
Here's a more detailed explanation of what I'm proposing:

## Source

```
src
├── SUMMARY.md
├── foo
│   ├── alpha.md
│   ├── beta.md
│   └── readme.md
├── bar
│   ├── delta.md
│   ├── gamma.md
│   └── readme.md
└── readme.md
```

## Current behaviour

```
book
├── foo
│   ├── alpha.html
│   ├── beta.html
│   └── index.html
├── bar
│   ├── delta.html
│   ├── gamma.html
│   └── index.html
└── index.html
```

## Proposed behaviour

```
book
├── foo
│   ├── alpha
│   │   └── index.html
│   ├── beta
│   │   └── index.html
│   └── index.html
├── bar
│   ├── delta
│   │   └── index.html
│   ├── gamma
│   │   └── index.html
│   └── index.html
└── index.html
```

---

This proposal is in line with the way [hugo](https://gohugo.io), [eleventy](https://www.11ty.io), and AFAIK even [jekyll](https://jekyllrb.com) work.

I suppose this could also take the form of an optional preprocessor which would create directories for every non-index markdown file and move them into the directory as `index.md` or `readme.md`.

**Comment by sjkim04** [comment]
I agree with this. While there are permalink options in most webpage renderers, mdBook does not. I will open another issue for this.
